src/blog/understanding-ecmascript-part-2.md

```diff
@@ -0,0 +1,358 @@
+
+We'll take a deeper look into the grammar rules in a later episode, let's keep it simple for now! In particular, we can ignore the subscripts (`Yield`, `Await` and so on) in the productions for this episode.
+
+The following productions describe how a `MemberExpression` looks like:
+
+> [`MemberExpression :`](https://tc39.es/ecma262/#prod-MemberExpression)
+>
+> `PrimaryExpression`
+> `MemberExpression [ Expression ]`
+> `MemberExpression . IdentifierName`
+> `MemberExpression TemplateLiteral`
+> `SuperProperty`
+> `MetaProperty`
+> `new MemberExpression Arguments`
+
+Here we have 8 productions for `MemberExpression`. A `MemberExpression` can be just a `PrimaryExpression`. Alternatively, a `MemberExpression` can be constructed from another `MemberExpression` and `Expression` by piecing them together: `MemberExpression [ Expression ]`, for example `o2['foo']`. Or it can be `MemberExpression . IdentifierName`, for example `o2.foo` &mdash; this is the production relevant for our example.
+
+Runtime semantics for the production `MemberExpression : MemberExpression . IdentifierName` define the set of steps to take when evaluating it:
+
+:::ecmascript-algorithm
+> [Runtime Semantics: Evaluation for `MemberExpression : MemberExpression . IdentifierName`](https://tc39.es/ecma262/#sec-property-accessors-runtime-semantics-evaluation)
+>
+> 1. Let `baseReference` be the result of evaluating `MemberExpression`.
+> 1. Let `baseValue` be `? GetValue(baseReference)`.
+> 1. If the code matched by this `MemberExpression` is strict mode code, let `strict` be `true`; else let `strict` be `false`.
+> 1. Return `? EvaluatePropertyAccessWithIdentifierKey(baseValue, IdentifierName, strict)`.
+
+The algorithm delegates to the abstract operation `EvaluatePropertyAccessWithIdentifierKey`, so we need to read it too:
+
+:::ecmascript-algorithm
+> [`EvaluatePropertyAccessWithIdentifierKey( baseValue, identifierName, strict )`](https://tc39.es/ecma262/#sec-evaluate-property-access-with-identifier-key)
+>
+> The abstract operation `EvaluatePropertyAccessWithIdentifierKey` takes as arguments a value `baseValue`, a Parse Node `identifierName`, and a Boolean argument `strict`. It performs the following steps:
+>
+> 1. Assert: `identifierName` is an `IdentifierName`
+> 1. Let `bv` be `? RequireObjectCoercible(baseValue)`.
+> 1. Let `propertyNameString` be `StringValue` of `identifierName`.
+> 1. Return a value of type Reference whose base value component is `bv`, whose referenced name component is `propertyNameString`, and whose strict reference flag is `strict`.
+
+That is: `EvaluatePropertyAccessWithIdentifierKey` constructs a Reference which uses the provided `baseValue` as the base, the string value of `identifierName` as the property name, and `strict` as the strict mode flag.
+
+Eventually this Reference gets passed to `GetValue`. This is defined in several places in the spec, depending on how the Reference ends up being used.
+
+### Property access as a parameter
```

Comment:
I found the first section compelling and cohesive: let's take a behavior we all know and thoroughly trace through how it's specified. However, I'm missing the transition to these next two sections. It feels a bit abrupt.

Is the idea to highlight a few places where property gets can appear, and then work those out? Why were function arguments and RHS of assignments chosen as the examples?

I actually think this episode should end here with the grammar parts left for the next episode. The narrative is "let's learn how a feature is specified from the bottom up". Episode 1 was the spec mechanism underpinnings: the definitions you need to even navigate the spec. Episode 2 builds on Episode 1 to completely explain how, given an expression like `o.p`, how that expression is specified. Episode 3 can then focus on the grammar: now that we know how `o.p` works, how does the spec get a `o.p`? You can then flesh out a full, but very simple program like `x = o2.foo` or even just `o2.foo`. Then you can dig into the grammar productions a bit more, with its recursive cases and so on. WDYT?

Reply:
As discussed, moved part to an extra content, got rid of one of the "property access" subsections.